csharp/Worker/Microsoft.Spark.CSharp/Worker.cs

```diff
@@ -455,185 +435,91 @@ private static long ToUnixTime(DateTime dt)
+                watch.Stop();
+                if (messageLength > 0 || messageLength == (int)SpecialLengths.NULL || mode == SerializedMode.Pair)
                 {
-                    items = GetNext(messageLength);
-                    Debug.Assert(items != null);
-                    Debug.Assert(items.Any());
-                    pos = 0;
-                    hasNext = true;
-                }
-                else
-                {
-                    //unexpected behavior. 
-                    //Try moving on to the next item. This might throw exception but
-                    //it might also fetch the next items successfully
-                    //So it is better not to throw exception right way
-                    return MoveNext();
-                }
-            }
-
-            watch.Stop();
-            return hasNext;
-        }
-
-        public object Current
-        {
-            get
-            {
-                int currPos = pos;
-                pos++;
-                return items[currPos];
-            }
-        }
-
-        public void Reset()
-        {
-            throw new NotImplementedException();
-        }
-
-        public IEnumerator GetEnumerator()
-        {
-            return this;
-        }
-
-        public void LogStatistic()
-        {
-            logger.LogInfo(string.Format("total elapsed time: {0}", watch.ElapsedMilliseconds));
-        }
-
-        private object[] GetNext(int messageLength)
-        {
-            object[] result = null;
-            switch ((SerializedMode)Enum.Parse(typeof(SerializedMode), deserializedMode))
-            {
-                case SerializedMode.String:
+                    watch.Start();
+                    byte[] buffer = messageLength > 0 ? SerDe.ReadBytes(inputStream, messageLength) : null;
+                    watch.Stop();
+                    switch (mode)
                     {
-                        result = new object[1];
-                        if (messageLength > 0)
-                        {
-                            byte[] buffer = SerDe.ReadBytes(inputStream, messageLength);
-                            if (buffer == null)
+                        case SerializedMode.String:
                             {
-                                logger.LogDebug("Buffer is null. Message length is {0}", messageLength);
+                                yield return buffer == null ? null : SerDe.ToString(buffer);
+                                break;
                             }
-                            result[0] = SerDe.ToString(buffer);
-                        }
-                        else
-                        {
-                            result[0] = null;
-                        }
-                        break;
-                    }
 
-                case SerializedMode.Row:
-                    {
-                        Debug.Assert(messageLength > 0);
```

Comment:
I think we should keep Debug.Assert() here, to ensure the logging/assertion characteristics as the same as before, unless we deem this assert() is completely unnecessary.